hibernate-core/src/test/java/org/hibernate/test/cache/ByteCodeEnhancedImmutableReferenceCacheTest.java

```diff
@@ -91,6 +97,132 @@ public void testUseOfDirectReferencesInCache() throws Exception {
+		assertFalse( s2.contains( myReferenceData ) );
+		assertFalse( s2.contains( myOtherReferenceData ) );
+
+		assertNull( ( (SharedSessionContractImplementor) s2 ).getPersistenceContext().getEntry( myReferenceData ) );
+		assertNull( ( (SharedSessionContractImplementor) s2 ).getPersistenceContext().getEntry( myOtherReferenceData ) );
+
+		// evict should do nothing, since p is not associated with s2
+		s2.evict( myReferenceData );
+		s2.evict( myOtherReferenceData );
+
+		assertSame( myReferenceData, s2.get( MyEnhancedReferenceData.class, myReferenceData.getId() ) );
+		assertSame( myOtherReferenceData, s2.get( MyEnhancedReferenceData.class, myOtherReferenceData.getId() ) );
+
+		assertTrue( s2.contains( myReferenceData ) );
+		assertTrue( s2.contains( myOtherReferenceData ) );
+
+		// still associated with s1
+		assertTrue( s1.contains( myReferenceData ) );
+		assertTrue( s1.contains( myOtherReferenceData ) );
+
+		s2.evict( myReferenceData );
+		s2.evict( myOtherReferenceData );
+
+		assertFalse( s2.contains( myReferenceData ) );
+		assertFalse( s2.contains( myOtherReferenceData ) );
+
+		s2.getTransaction().commit();
+		s2.close();
+
+		// still associated with s1
+		assertTrue( s1.contains( myReferenceData ) );
+		assertTrue( s1.contains( myOtherReferenceData ) );
+
+		s1.clear();
+
+		assertFalse( s1.contains( myReferenceData ) );
+		assertFalse( s1.contains( myOtherReferenceData ) );
+
+		s1.close();
+
+		// EntityEntry should still be set
+		assertNotNull( myReferenceData.$$_hibernate_getEntityEntry() );
+		assertNotNull( myOtherReferenceData.$$_hibernate_getEntityEntry() );
+
+		// load them into 2 sessions
+		s1 = openSession();
+		s1.getTransaction().begin();
+		assertSame( myReferenceData, s1.get( MyEnhancedReferenceData.class, myReferenceData.getId() ) );
+		assertSame( myOtherReferenceData, s1.get( MyEnhancedReferenceData.class, myOtherReferenceData.getId() ) );
+
+		s2 = openSession();
+		s2.getTransaction().begin();
+		assertSame( myReferenceData, s2.get( MyEnhancedReferenceData.class, myReferenceData.getId() ) );
+		assertSame( myOtherReferenceData, s2.get( MyEnhancedReferenceData.class, myOtherReferenceData.getId() ) );
+
+		assertEquals( Status.READ_ONLY, myReferenceData.$$_hibernate_getEntityEntry().getStatus() );
+		assertEquals( Status.READ_ONLY, myOtherReferenceData.$$_hibernate_getEntityEntry().getStatus() );
+
+		// delete myReferenceData from s1
+		s1.delete( myReferenceData );
+
+		assertEquals( Status.DELETED, myReferenceData.$$_hibernate_getEntityEntry().getStatus() );
+		assertEquals( Status.READ_ONLY, myOtherReferenceData.$$_hibernate_getEntityEntry().getStatus() );
+
+		// delete myOtherReferenceData from s2
+		s2.delete( myOtherReferenceData );
+
+		assertEquals( Status.DELETED, myReferenceData.$$_hibernate_getEntityEntry().getStatus() );
+		assertEquals( Status.DELETED, myOtherReferenceData.$$_hibernate_getEntityEntry().getStatus() );
+
+		s1.getTransaction().commit();
+		s1.close();
+
```

Comment:
s2 still contains myReferenceData. Now it's status is Status.GONE. I suspect this could cause problems.

@sebersole, WDYT?